fs_test.go

```diff
@@ -44,6 +45,186 @@ func TestMain(m *testing.M) {
+		fixtures := []struct {
+			desc   string
+			offset int64
+			whence int
+		}{
+			{"seek before beginning", -1, io.SeekCurrent},
+			{"seek after end", int64(len(content)) + 1, io.SeekStart},
+			{"seek exactly to end", 0, io.SeekEnd},
+		}
+
+		for _, f := range fixtures {
+			f := f
+			t.Run(f.desc, func(t *testing.T) {
+				testFs := s3fs.New(s3cl, *bucket)
+				data, err := testFs.Open(testFile)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				_, err = data.(io.Seeker).Seek(f.offset, f.whence)
+				if err == nil {
+					t.Fatal(err)
+				}
+
+				errorType, ok := err.(awserr.RequestFailure)
+
+				if !ok {
+					t.Errorf("Expected awserr.RequestFailure, got %v", errorType)
+				}
+
+			})
+		}
+	})
+
+	t.Run("seek from other starting position", func(t *testing.T) {
+		fixtures := []struct {
+			desc          string
+			initialOffset int
+			offset        int64
+			whence        int
+			expected      int64
+		}{
+			{"whence SeekStart", 3, 2, io.SeekStart, 2},
+			{"whence SeekCurrent", 3, 3, io.SeekCurrent, 6},
+			{"whence SeekEnd", 3, -1, io.SeekEnd, int64(len(content)) - 1},
+		}
+
+		for _, f := range fixtures {
+			f := f
+			t.Run(f.desc, func(t *testing.T) {
+				testFs := s3fs.New(s3cl, *bucket)
+				data, err := testFs.Open(testFile)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				readBuffer := make([]byte, f.initialOffset)
+				readBytes, err := data.Read(readBuffer)
+				if err != nil {
+					t.Fatal(err)
+				}
+				if readBytes != f.initialOffset {
+					t.Errorf("Read failed during test setup")
+				}
+
+				actual, err := data.(io.Seeker).Seek(f.offset, f.whence)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				if actual != f.expected {
+					t.Errorf("Expected %d, got %d", f.expected, actual)
+				}
+
+			})
+		}
+	})
+
+	t.Run("seek then read", func(t *testing.T) {
+		fixtures := []struct {
+			desc      string
+			readBytes int
+			offset    int64
+			whence    int
+			expected  []byte
+		}{
+			{"whence SeekStart", 3, 2, io.SeekStart, content[2:5]},
+			{"whence SeekStart overflow", 3, 5, io.SeekStart, content[5:7]},
+			{"whence SeekCurrent", 3, 4, io.SeekCurrent, content[4:7]},
+			{"whence SeekEnd", 3, -3, io.SeekEnd, content[len(content)-3:]},
+		}
+
+		for _, f := range fixtures {
+			f := f
+			t.Run(f.desc, func(t *testing.T) {
+				testFs := s3fs.New(s3cl, *bucket)
+				data, err := testFs.Open(testFile)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				_, err = data.(io.Seeker).Seek(f.offset, f.whence)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				readBuffer := make([]byte, f.readBytes)
+				readBytes, err := data.Read(readBuffer)
+				if readBytes != len(f.expected) {
+					t.Errorf("Read returned unexpected number of bytes")
+				}
+				if err != nil && err != io.EOF {
+					t.Fatal(err)
+				}
+				if bytes.Compare(readBuffer[:readBytes], f.expected) != 0 {
+					t.Errorf("Expected %s, got %s", string(f.expected), string(readBuffer))
+				}
```

Comment:
I am paranoid so let's parallel this with `bytes.Reader` to be sure that our seeker returns the same

```go
				readSeekers := []struct {
					desc   string
					seeker io.ReadSeeker
				}{
					{desc: "bytes reader", seeker: bytes.NewReader(content)},
					{desc: "file", seeker: data.(io.ReadSeeker)},
				}

				for _, rs := range readSeekers {
					t.Run(f.desc, func(t *testing.T) {
						_, err = rs.seeker.Seek(f.offset, f.whence)
						if err != nil {
							t.Fatal(err)
						}

						readBuffer := make([]byte, f.readBytes)
						readBytes, err := rs.seeker.Read(readBuffer)
						if readBytes != len(f.expected) {
							t.Errorf("Read returned unexpected number of bytes")
						}
						if err != nil && err != io.EOF {
							t.Fatal(err)
						}
						if bytes.Compare(readBuffer[:readBytes], f.expected) != 0 {
							t.Errorf("Expected %s, got %s", string(f.expected), string(readBuffer))
						}
					})
				}

```

Reply:
I tried this, it does not work immediately as the `Read()` behaviour is different between the two:
The bytes.Reader returns io.EOF on a Read() call only when there is nothing left to read and there are no bytes being read.
The ReadCloser that comes back from s3 returns io.EOF on a Read() call if the offset is past the last readable byte after filling the buffer.

To illustrate this with an example:
8 bytes left to read, buffer of size 16:
bytes.Reader reads 8 bytes and returns `8, nil` on the first Read call and `0, io.EOF` on further ones.
"Our" ReadCloser reads 8 bytes and returns `8, io.EOF` on the first call already.

According to this, both are acceptable behaviours:
https://pkg.go.dev/io#Reader
> An instance of this general case is that a Reader returning a non-zero number of bytes at the end of the input stream may return either err == EOF or err == nil. The next Read should return 0, EOF.

As far as I can see, there are 4 ways to solve this:
1. Adapt the `Read()` method in file.go to mimic bytes.Reader behaviour.
2. Don't do the parallel tests for cases where one of the two readers would return io.EOF and the other wouldn't.
3. Add an additional `Read()` call to the tests that read till end of file, to compare the error of this second call, because that call should return `0, io.EOF` for both readers. Ignore differences in errors on the first call, if the error is one of the two valid ones (nil or io.EOF).
4. Distinguish between the two readers in the tests if we expect an EOF, which would be very ugly.

I have chosen 3. for now

(This is by the way not a thing that has changed with the implementation of the Seeker interface - the behaviour of the two readers was already different before)

Reply:
our test should not be affected by the behavior of the reader (for as long as it's correct)

wouldn't using `io.CopyN` solve this? I implemented this and all tests are passing

```go
				readSeekers := []struct {
					desc   string
					seeker io.ReadSeeker
				}{
					{desc: "bytes reader", seeker: bytes.NewReader(content)},
					{desc: "file", seeker: data.(io.ReadSeeker)},
				}

				for _, rs := range readSeekers {
					t.Run(f.desc, func(t *testing.T) {
						_, err = rs.seeker.Seek(f.offset, f.whence)
						if err != nil {
							t.Fatal(err)
						}

						var buf bytes.Buffer
						_, err := io.CopyN(&buf, rs.seeker, int64(f.readBytes))

						if err != nil {
							t.Fatal(err)
						}

						if buf.String() != string(f.expected) {
							t.Errorf("Expected %s, got %s", string(f.expected), buf.String())
						}
					})
				}
```

Reply:
Using this now.